Update mocha, chai for testomat reporting internal

achakko opened the pull request:
### Description:

This is a pull request to update mocha, chai version for all tests, thereby to help enable CI integration with [testomat.io](http://testomat.io/) to understand reporting on tests including flakiness

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

DavertMik committed:
[WIP] Added testomatio reporter for ui tests (#21166)

* added testomatio reporter for ui tests

* removed api key from code

---------

Co-authored-by: achakko <141092099+achakko@users.noreply.github.com>

sgiehl reviewed (changes requested):
Seems you have commited a couple changes to the submodules. This for sure shouldn't be included in this PR and currently causes a lot tests to fail.
You may need to revert those changes. If you need help with git or how to handle submodules feel free to get in touch with me or another dev on slack

achakko commented:
Hi @sgiehl @michalkleiner I rebased to remove the commit that was not required. Thanks.

achakko committed:
Merge branch '5.x-dev' into dev-17222-testomat-internal

mneudert commented on tests/lib/screenshot-testing/run-tests.js:
```diff
@@ -43,8 +43,8 @@ async function main() {
         ui: 'bdd',
-        reporter: config.reporter,
         bail: false,
+        reporter: 'node_modules/@testomatio/reporter/lib/adapter/mocha',
```
Should the only change in this PR be the version bump for mocha?

To properly activate the integration we need some more changes:

- some multi reporter plugin, otherwise we lose the regular spec output both locally and in CI (the [testomat examples](https://github.com/testomatio/examples/tree/master/mocha-ts-multi-reporters) have an example)
- If the reporter does not silently do nothing if the API key is missing we also need some changes so the tests run without the API key. The test run here seems to have not uploaded anything but also not failed, so it probably is good the way it is
- Having the API key configured in the UI test environment, should work at the job level here withot having to modify the testing action

achakko replied:
I heard back from testomatio, here's updates to your feedback

Mocha update is required because in mocha 8 they introduced constants for passed/failed tests. We rely on these constants and probably this is the only issue here. Even, we may probably support old mocha versions it's a dead end path as software should be constantly updated.

1. Our reporter should do some output in a spec-like manner, so no multi-report should be needed (see screenshot)
2. Yes, reporter won't work without API key. I think we may remove that requirement, as mocha reporter still can print output data without a key. We can update it!
3. Yes, you can pass API key to job action as an environment variable and that will enable reporting
![image (4)](https://github.com/matomo-org/matomo/assets/141092099/5d8cdf6e-600d-4c5b-8f4a-e17e1dd64f4e)

achakko replied:
For # 2, that should be done on testomatio side. 

For # 3, @matomo-org/core-team do you know how to pass the testomatio environment variable that Thomas created into the Github actions job? I think we use custom actions to run tests, can a custom action load TESTOMATIO secret and set it as environment variable?

bx80 replied:
If you take a look at some of the existing github actions they [retrieve secrets from environment variables](https://github.com/matomo-org/matomo/blob/a89e26dd7e05b56ce453796e86f81637b7a9808a/.github/workflows/matomo-tests.yml#L61).  `${{ secrets.TESTOMATIO }}` should get the API key in the action and then you should be able to pass it to mocha using a .env file and finally access it in JavaScript with ```process.env.TESTOMATIO```

So in the github action do something like:
```
run: |
    touch .env
    echo TESTOMATIO_API_KEY=${{ secrets.TESTOMATIO }} >> .env
```
then in `run-tests.js` you should be able to do:
```
let api_key = process.env.TESTOMATIO;
```

mneudert replied:
Regarding the output:

Yeah, the adapter does some output, but, from [what I can see](https://github.com/testomatio/reporter/blob/master/lib/adapter/mocha.js), it is subpar to the regular spec output (missing indentation, missing describe blocks, no test runtimes). However, I haven't run that one locally, so please take that with a big grain of salt, there might be some magic happening that merges everything together into the "regular spec output" style.

Also two days ago there was a [change](https://github.com/testomatio/reporter/commit/2f92d655f2ff8a746327151208a2d919c9b953ea) to the reporter, so it should now always output something, even if the API key is missing. Without that change we will probably still need a solution for non-CI environments, e.g. a fallback to the regular spec reporter if no API key is found, but multi-reporter should be unnecessary and only create a complete output mess.

achakko replied:
> If you take a look at some of the existing github actions they [retrieve secrets from environment variables](https://github.com/matomo-org/matomo/blob/a89e26dd7e05b56ce453796e86f81637b7a9808a/.github/workflows/matomo-tests.yml#L61). `${{ secrets.TESTOMATIO }}` should get the API key in the action and then you should be able to pass it to mocha using a .env file and finally access it in JavaScript with `process.env.TESTOMATIO`
> 
> So in the github action do something like:
> 
> ```
> run: |
>     touch .env
>     echo TESTOMATIO_API_KEY=${{ secrets.TESTOMATIO }} >> .env
> ```
> 
> then in `run-tests.js` you should be able to do:
> 
> ```
> let api_key = process.env.TESTOMATIO;
> ```

Added the environment variable to matomo-tests.yml , tests ran and failed (expected I think), but I don't see the reporting in testomat. I am going to follow up with testomat for this one.

sgiehl replied:
> Added the environment variable to matomo-tests.yml

That file wasn't changed in this PR, so not sure what you tried to do

achakko replied:
Hi @sgiehl I was able to clean up my local branches and made sure my commits made it in this time :) Thanks for clearing that up. Now that the change is in, somehow I managed to not run all the required tests. I think I did not update the actions correct in matomo-tests.xml. Going to review with the testomat team.

mneudert commented on tests/lib/screenshot-testing/package-lock.json:
```diff
@@ -1,2724 +1,1579 @@
   "name": "matomo-screenshot-testing",
   "version": "1.0.0",
-  "lockfileVersion": 2,
+  "lockfileVersion": 1,
```
Can you regenerate the lockfile with a more recent version of `npm`? Anything from `v7` onwards should generate the expected version.

achakko committed:
Merge branch '5.x-dev' into dev-17222-testomat-internal

achakko committed:
Merge branch '5.x-dev' into dev-17222-testomat-internal

achakko committed:
Added testomatio environment variable to matomo-tests.yml

achakko committed:
Added testomatio environment variable to matomo-tests.yml

achakko commented:
@matomo-org/core-reviewers This reporting seems to be working now. The CI job ran tests and reported the UI test run results into testomat with a decent stack trace for failed tests as well. Runs can be accessed here. Once we have more data, will get meaningful results in Analytics. 
https://app.testomat.io/projects/core-ui-tests/runs
![image](https://github.com/matomo-org/matomo/assets/141092099/7314178b-cda4-45e2-a030-bfbe92f35056)

bx80 reviewed (changes requested):
The test tomato reporting is looking great :+1:

Is there any reason the UI test output log has lots of `[TESTOMATIO] Warning: Test could not be matched (404) should be possible to publish new version` messages? This will make  it a bit harder to check the logs, it also looks like colour highlighting has also disappeared which is helpful to quickly identify the failed tests.

achakko committed:
Added testomatio environment variable to matomo-tests.yml

achakko committed:
Added testomatio environment variable to matomo-tests.yml

achakko committed:
Added mocha-multi-reporter to package.jso to enable mocha reporting for standard spec and testomat, also added to config.

achakko committed:
Added mocha-multi-reporter to package.jso to enable mocha reporting for standard spec and testomat, also added to config.

achakko committed:
added testomatio reporter in addition regular spec

achakko committed:
added testomatio reporter in addition regular spec

achakko committed:
Updated one test to ensure it is reported into testomat correctly.

achakko committed:
Merge branch '5.x-dev' into dev-17222-testomat-internal

achakko committed:
Updated one test to ensure it is reported into testomat correctly.

achakko committed:
Merge remote-tracking branch 'origin/dev-17222-testomat-internal' into dev-17222-testomat-internal

achakko committed:
Added comments for multiple mocha reporters

achakko committed:
Adding nyc mocha to report test coverage

achakko commented:
> The test tomato reporting is looking great 👍
> 
> Is there any reason the UI test output log has lots of `[TESTOMATIO] Warning: Test could not be matched (404) should be possible to publish new version` messages? This will make it a bit harder to check the logs, it also looks like colour highlighting has also disappeared which is helpful to quickly identify the failed tests.

Hi @bx80 I was able to remove the warnings. See [one of the recent runs here](https://github.com/matomo-org/matomo/actions/runs/6791191118/job/18462192513?pr=21400).  Do you think the absence of the colour highlighting is a deal breaker?

michalkleiner commented on .github/workflows/matomo-tests.yml:
```diff
@@ -110,10 +110,11 @@ jobs:
           redis-service: true
           artifacts-pass: ${{ secrets.ARTIFACTS_PASS }}
-          upload-artifacts: true
+          upload-artifacts: true
```
please keep new lines at the end of files

michalkleiner commented on tests/UI/config.dist.js:
```diff
@@ -38,12 +38,12 @@ exports.mocha = 'mocha-3.1.2';
 
 /**
- * The mocha reporter to use.
+ * The multiple mocha reporters to use.
```
```suggestion
 * Mocha reporters to use (can be multiple delimited by a comma).
```

mneudert commented on package.json:
```diff
@@ -71,6 +72,7 @@
     "husky": "^8.0.2",
     "less": "^4.1.2",
     "less-loader": "^7.3.0",
+    "mocha-multi-reporters": "^1.5.1",
```
This is the wrong file to add your new packages to, everything should go into `tests/lib/screenshot-testing/package.json`.

achakko replied:
Thank you @mneudert for not only pointing this out but also with the code to enable mocha-multi-reports correctly. Updated run-tests.js and config.dist.js based on your review.

mneudert commented on tests/UI/config.dist.js:
```diff
@@ -38,12 +38,12 @@ exports.mocha = 'mocha-3.1.2';
+ * The multiple mocha reporters to use.
  */
-exports.reporter = "spec";
+exports.reporter = 'spec, @testomatio/reporter/lib/adapter/mocha.js';
```
This configuration has no effect until `tests/lib/screenshot-testing/run-tests.js` is configured to use the previous `reporter: config.reporter,` instead of the fixed testomatio reporter (or rather the new multi-reporter as that is probably what you want here).

achakko committed:
Updated multiple mocha reporter comment and added mocha-multi-reporter dev dependency to screenshot-testing/package.json

achakko committed:
Merge branch '5.x-dev' into dev-17222-testomat-internal

achakko committed:
Added mocha-multi-reporter dev dependency to screenshot-testing/package.json

achakko committed:
Merge remote-tracking branch 'origin/dev-17222-testomat-internal' into dev-17222-testomat-internal

achakko committed:
Added mocha-multi-reporter dev dependency to screenshot-testing/package.json

achakko committed:
Added mocha-multi-reporter dev dependency to screenshot-testing/package.json

achakko committed:
Added mocha-multi-reporter dev dependency to screenshot-testing/package.json

achakko committed:
deleting .npmrc file

achakko committed:
updated lockfile

achakko committed:
Configured mocha-multi-reporter in config and run-tests files

achakko committed:
typo in reporters

achakko committed:
Trying to force mocha colours

achakko committed:
Trying to force mocha colours

achakko committed:
removed deprecated useColors method and kept color method

achakko commented:
> > The test tomato reporting is looking great 👍
> > Is there any reason the UI test output log has lots of `[TESTOMATIO] Warning: Test could not be matched (404) should be possible to publish new version` messages? This will make it a bit harder to check the logs, it also looks like colour highlighting has also disappeared which is helpful to quickly identify the failed tests.
> 
> Hi @bx80 I was able to remove the warnings. See [one of the recent runs here](https://github.com/matomo-org/matomo/actions/runs/6791191118/job/18462192513?pr=21400). Do you think the absence of the colour highlighting is a deal breaker?

Added color = 1 and removed the deprecated Usecolors method. Seems to be working! We have a summary of any failed just above the Error message with colors ofcourse. Good to approve?
![image](https://github.com/matomo-org/matomo/assets/141092099/af720bc1-5024-48ba-b0c0-106ce039ae7c)

bx80 reviewed (approved):
Thanks @achakko, much easier to read now :+1: 
Might be good to get additional approvals for this one, as I've not been following all the other points raised.

sgiehl commented on tests/lib/screenshot-testing/package.json:
```diff
@@ -9,10 +9,14 @@
     "url-parse": "^1.5.3"
+  },
+  "devDependencies": {
+    "@testomatio/reporter": "^1.0.9",
+    "mocha-multi-reporters": "^1.5.1"
```
Actually there is no need to define those as dev dependencies only, as actually the whole part is used in development only.

achakko replied:
Removed these from dev dependencies and updated package.json. Thanks @sgiehl .

sgiehl commented on tests/UI/config.dist.js:
```diff
@@ -38,12 +38,18 @@ exports.mocha = 'mocha-3.1.2';
+ * Mocha reporters to use (can be multiple delimited by a comma).
  */
-exports.reporter = "spec";
+exports.reporter = 'mocha-multi-reporters';
+exports.reporterOptions = {
+  reporterEnabled: 'spec, @testomatio/reporter/lib/adapter/mocha.js',
+  testomatioReporterLibAdapterMochaJsReporterOptions: {
+    apiKey: process.env.TESTOMATIO
+  }
+};
```
Shouldn't we only set this if `process.env.TESTOMATIO` has a value. Otherwise all plugin tests or even when running tests locally will try to use the TESTOMATIO reporter (with an empty api key).

achakko replied:
Added condition if `process.env.TESTOMATIO` has value, run testomat reporting, else run regular spec. Thanks @sgiehl

achakko commented:
The issue with the double reporting in testomat.io is logged here. I will follow up for resolution. https://github.com/testomatio/reporter/issues/245

achakko committed:
Merge branch '5.x-dev' into dev-17222-testomat-internal

achakko committed:
Updated dependencies in package.json and updated condition to run spec tests locally and testomat in CI

achakko committed:
Merge remote-tracking branch 'origin/dev-17222-testomat-internal' into dev-17222-testomat-internal

achakko committed:
Merge branch '5.x-dev' into dev-17222-testomat-internal

michalkleiner commented on tests/lib/screenshot-testing/run-tests.js:
```diff
@@ -43,9 +43,10 @@ async function main() {
         bail: false,
-        useColors: true,
+        reporter: config.reporter,
+        reporterOptions: config.reporterOptions,
```
with the newly added config condition the reporterOptions may be undefined from there
```suggestion
        reporterOptions: config.reporterOptions || {},
```

mneudert replied:
The options will now be always exported from the config instead of having to check if the exports are complete.

mneudert committed:
Update screenshot-testing lockfile

mneudert committed:
Always export reporterOptions from UI config

sgiehl committed:
revert changes to package-lock.json

sgiehl committed:
Merge branch '5.x-dev' into dev-17222-testomat-internal

sgiehl reviewed (approved):
This overall looks good to me now. Don't have access to the testomatio dashboard, so can't say if it's working on that side.
